src/liana/method/sp/_inflow.py

```diff
@@ -0,0 +1,345 @@
+        w = _handle_connectivity(adata=adata, connectivity_key=connectivity_key)
+
+        k = ct.shape[1]           # number of cell types
+        m = x_mat.shape[1]       # number of LR pairs
+
+        # Initialize empty list to hold each (cell x ligand) matrix per celltype
+        ls_list = []
+
+        # Loop over each cell type column in `ct` (a sparse binary matrix)
+        for i in range(ct.shape[1]):
+            # Slice the indicator column for one cell type: (n_cells, 1)
+            ct_i = ct[:, i]
+
+            # Elementwise multiply x_mat (ligand expr) with cell type indicator
+            # This will zero out cells not in this cell type
+            ls_i = x_mat.multiply(ct_i)
+
+            ls_list.append(ls_i)
+
+        # Horizontally stack to simulate (n_cells, n_celltypes * n_ligands)
+        ls = hstack(ls_list)  # shape: (n_cells, k * m)
+
+        # Min-max transform the ligand * celltype data & apply spatial weighting
+        if not isinstance(ls, csr_matrix):
+            ls = csr_matrix(ls)
+
+        # Transform ligand matrix
+        ls = self._transform(ls, x_transform, **x_transform_kwargs)
+
+        wls = w.dot(ls)
+
+        # Normalize by row sums (avoid division by zero)
+        row_sums = np.asarray(w.sum(axis=1)).flatten()
+        row_sums[row_sums == 0] = 1.0  # avoid division by zero
+        inv_row_sums = 1.0 / row_sums
+        wls = wls.multiply(inv_row_sums.reshape(-1, 1))  # still sparse
+
+        # Clean NaNs in sparse matrix (if any)
+        wls.data[np.isnan(wls.data)] = 0
+
+        # Transform receptor matrix
+        r = self._transform(y_mat, y_transform, **y_transform_kwargs)
+
+        # Ensure r is sparse and repeat across cell types
+        if not isinstance(r, csr_matrix):
+            r = csr_matrix(r)
+        ri = hstack([r] * k)  # replicate across k cell types - changed
+
+        # Sparse elementwise multiplication
+        xy_mat = wls.multiply(ri)  # both are sparse
+
+
+        # Create .var index: each column is "cell_type ^ interaction_name"
+        var = pd.DataFrame(
+            index=(
+                np.repeat(ct_labels.astype(str), m) +
+                xy_sep +
+                np.tile(xy_stats['interaction'].astype(str), k)
+            )
+        )
+
+        # Construct the output AnnData
+        lrdata = sc.AnnData(
+            X=csr_matrix(xy_mat),
+            var=var,
+            obs=adata.obs,
+            uns=adata.uns,
+            obsm=adata.obsm,
+            varm=adata.varm,
+            obsp=adata.obsp
+        )
+
+        # Drop non-variable features
+        _, var = mean_variance_axis(lrdata.X, axis=0)
+        lrdata = lrdata[:, var > 0]
+
+        X = lrdata.X.astype(float)
+        mean = np.asarray(X.mean(axis=0)).ravel()
+        mean_sq = np.asarray(X.power(2).mean(axis=0)).ravel()
+        var = mean_sq - mean**2
+        std = np.sqrt(var)
+        cv = std / (mean + 1e-12)
+
+        gini_vals = []
```

Comment:
This seems to be easily replacable with a vector operation (as other fiunctions) rather than a for loop; my assumption is that this is what made inflow much slower than before.